<h1 align="center">Python performance exercises</h1>

## Python best practices exercises

### Exercise 1

considering the following function for concatenating list strings with deliter.

In [3]:
def ft_concatenate(l_strings, d):
    """concatenate list of strings into one string separated by delimeter"""
    res = l_strings[0]
    for e in l_strings[1:]:
        res = res + d + e
    return res

- profile the function and identify the bottlenecks.
- improve speed up of the function
*Hint: you may need to look to the string functions in python documentation*

In [1]:
%load_ext line_profiler

In [2]:
l_strings = ["Hello", "Sir", "How are you ?"]
d = "-"

In [7]:
%timeit ft_concatenate(l_strings, d)

404 ns ± 5.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [4]:
%lprun -f ft_concatenate ft_concatenate(l_strings, d)

### Improving speed of the function

In [5]:
def opt_ft_concatenate(l_strings, d):
    return d.join(l_strings)

In [8]:
%timeit opt_ft_concatenate(l_strings, d)

176 ns ± 2.55 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [6]:
%lprun -f opt_ft_concatenate opt_ft_concatenate(l_strings, d)

### Exercise 2

In this exercise you will solve the following problem using two methods bruteforce mehode, and fast method.

**Problem:** You are given a list of n integers, and your task is to calculate the number of distinct values in the list.

**Example**
- Input:
5
2 3 2 2 3

- Output:
2

**Implement the following methodes:**

1. **bruteforce mehode:** create an empty list and start adding items for the given list without adding the previous item add, at the end the result list will contain unique values, print lenght of the list and you are done. 
2. **fast method** think of using Set data structure.

- time the two methods, what do you think?

In [9]:
# bruteforce fast method
def distinct_values(l_integers):
    unique_val = []
    for val in l_integers:
        if val not in unique_val:
            unique_val += [val]

    return len(unique_val)

In [10]:
# write fast method
def fast_distinct_values(l_integers):
    
    # List to set
    l_integers_set = set(l_integers)
    
    return len(l_integers_set)

In [12]:
# Creat two random lists of numbers for testing
import numpy as np

# Generate 7 random numbers between 0 and 3
l1 = np.random.randint(4, size=7)
l2 = np.random.randint(4, size=7)

# time the two methods
print("Time for Bruteforce method")
%timeit distinct_values(l1)

print("\n")

print("Time for fast method")
%timeit fast_distinct_values(l1)

Time for Bruteforce method
2.12 µs ± 76.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


Time for fast method
1.6 µs ± 44 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


## Cython exercises

### Exercise 1

1. load the cython extension.

In [13]:
%load_ext cython

2. Condidering the following polynome function:

In [16]:
def poly(a,b):
    return 10.5 * a + 3 * (b**2)

- Create an equivalent Cython function of `poly` with name `poly_cy` without any cython improvement, just make its cell a cython cell.

In [17]:
%%cython -a
def poly_cy(a,b):
    return 10.5 * a + 3 * (b**2)

3. time the performance of Python and Cython version of the function, what is the factor of speed up here between the two verions.

In [18]:
# write your code here
print("Time for poly")
%timeit poly(2, 1)

print("\n")

print("Time for poly_cy")
%timeit poly_cy(2, 1)

Time for poly
315 ns ± 12.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


Time for poly_cy
286 ns ± 1.94 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [19]:
factor = 315/286
print("The factor of speed is : ", factor)

The factor of speed is :  1.1013986013986015


**This means that poly_cy is speedier than poly**

4. Now lets work on another examples using loop.
    - rewrite the same function below fib that calculate fibonacci series using cython, but now try to add type for the variables used inside it, add a prefix `_cy` to your new cython function.

In [26]:
def fib(n):
    a, b = 1, 1
    for i in range(n):
        a, b = a + b, a

    return a

In [25]:
%%cython -a
def fib_cy(int n):
    
    cdef int a = 1
    cdef int b = 1
    
    for i in range(n):
        a, b = a + b, a

    return a

- time the two function for fibonacci series, with n = 20, what is the factor of speed now, What do you think?

In [27]:
# write your code here
print("Time for fib")
%timeit fib(20)

print("\n")

print("Time for fib_cy")
%timeit fib_cy(20)

Time for fib
978 ns ± 8.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


Time for fib_cy
58.4 ns ± 0.0803 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [28]:
factor = 978/58.4
print("The factor of speed is : ", factor)

The factor of speed is :  16.746575342465754


**This shows us that fib_cy is 17 speedier than fib**

5. Recursive functions are functions that call themselves during their execution. Another interesting property of the Fibonacci series is that it can be written as a recursive function. That’s because each item depends on the values of other items (namely item n-1 and item n-2)

- Rewrite the fib function using recursion. Is it faster than the non-recursive version? Does Cythonizing it give even more of an advantage? 

In [29]:
def fib_rec(n):
    if n <= 1:
        return n
    else:
        return(fib_rec(n-1) + fib_rec(n-2))

In [30]:
%%cython -a
def fib_rec_cy(int n):
    if n <= 1:
        return n
    else:
        return(fib_rec_cy(n-1) + fib_rec_cy(n-2))

In [31]:
print("Time for fib_rec")
%timeit fib_rec(20)

print("\n")

print("Time for fib_rec_cy")
%timeit fib_rec_cy(20)

Time for fib_rec
2.49 ms ± 265 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Time for fib_rec_cy
468 µs ± 19.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


### Exercise 2

- Monte Carlo methods are a broad class of computational algorithms that rely on repeated random sampling to obtain numerical results. 
- One of the basic examples of getting started with the Monte Carlo algorithm is the estimation of Pi.

**Estimation of Pi**

- The idea is to simulate random (x, y) points in a 2-D plane with domain as a square of side 1 unit. 
- Imagine a circle inside the same domain with same diameter and inscribed into the square. 
- We then calculate the ratio of number points that lied inside the circle and total number of generated points. 
- Refer to the image below:

![demo](../data/MonteCarloPlot.png)

We know that area of the square is 1 unit sq while that of circle is $\pi \ast  (\frac{1}{2})^{2} = \frac{\pi}{4}$. Now for a very large number of generated points,

![demo](../data/MonteCarloCalc.png)


## The Algorithm

1. Initialize cile_points, square_points and interval to 0.
2. Generate random point x.
3. Generate random point y.
4. Calculate d = x*x + y*y.
5. If d <= 1, increment circle_points.
6. Increment square_points.
7. Increment interval.
8. If increment < NO_OF_ITERATIONS, repeat from 2.
9. Calculate pi = 4*(circle_points/square_points).
10. Terminate.

**Your mission:** time the function `monte_carlo_pi`, identify the bottlenecks and create a new version using cython functionality to speed up monte carlo simulation for PI, use 100,000 points and compare the speed up factor between python and cython, considering the following optimizations:
- add type for variables used.
- add type for the function
- use c rand function instead of python rand function.
 
*Hint: you can import function from C libraries using the following approach `from libc.<name of c library> cimport <library function name>`, replace the holders `<>` with the right identities for the current problem*

In [32]:
import random
def monte_carlo_pi(nsamples):
    pi = 0.
    
    cile_points, square_points = 0., 0.
    
    for i in range(nsamples):
        
        x = random.random()
        y = random.random()
        
        d = x*x + y*y
        if d <= 1:
            cile_points += 1
        
        square_points += 1
    
    pi = 4*(cile_points/square_points)
    return pi

In [33]:
%%cython -a
from libc.stdlib cimport rand, RAND_MAX

cpdef double monte_carlo_pi_cy(int nsamples):
    
    cdef double pi = 0.
    cdef double cile_points = 0.
    cdef double square_points = 0.
    
    cdef double x, y, d
    
    for i in range(nsamples):
        x = (rand()+1.0)/(RAND_MAX+2.0)
        y = (rand()+1.0)/(RAND_MAX+2.0)
        d = x*x + y*y
        if d <= 1:
            cile_points += 1
        
        square_points += 1
    
    pi = 4*(cile_points/square_points)
    return pi

In [34]:
print("Time for Monte Carlo Pi")
%timeit monte_carlo_pi(100000)
print("\n")
print("Time for Monte Carlo Pi with Cython")
%timeit monte_carlo_pi_cy(100000)

Time for Monte Carlo Pi
33.1 ms ± 569 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


Time for Monte Carlo Pi with Cython
3.55 ms ± 107 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [35]:
factor = 33.1/3.55
print("Factor of speed = ", factor)

Factor of speed =  9.323943661971832


**This gives an idea about the speed of cython. We can assume here that Cython is 10x faster than Python**

## Numba exercises

### Exercise 1

Previously we considered how to approximateby Monte Carlo.

- Use the same idea here, but make the code efficient using Numba.
- Compare speed with and without Numba when the sample size is large.

In [37]:
from numba import njit

In [38]:
@njit(fastmath=True)
def monte_carlo_pi_numba(nsamples):
    pi = 0.
    cile_points, square_points = 0., 0.
    
    for i in range(nsamples):
        
        x = random.random()
        y = random.random()
        
        d = x*x + y*y
        if d <= 1:
            cile_points += 1
        
        square_points += 1
    
    pi = 4*(cile_points/square_points)
    return pi

In [40]:
for n in [1000, 10000, 100000, 1000000]:
    print("Size = ", n)
    print("Time for Monte Carlo Pi")
    %timeit monte_carlo_pi(n)

    print("Time for Monte Carlo Pi with Numba")
    %timeit monte_carlo_pi_numba(100000)
    print("\n")

Size =  1000
Time for Monte Carlo Pi
329 µs ± 22.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Time for Monte Carlo Pi with Numba
1.02 ms ± 6.89 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


Size =  10000
Time for Monte Carlo Pi
3.17 ms ± 50.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Time for Monte Carlo Pi with Numba
1.01 ms ± 3.55 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


Size =  100000
Time for Monte Carlo Pi
31.7 ms ± 547 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Time for Monte Carlo Pi with Numba
1.09 ms ± 94.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


Size =  1000000
Time for Monte Carlo Pi
318 ms ± 4.42 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Time for Monte Carlo Pi with Numba
1.1 ms ± 82.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)




**We can see that more the size is large more Numbas is faster than simple Python code. Then the speed of Numba is approximately egal to 1.1 ms with N = 100000.**

## Pyccel Exercises

### Exercise 1

considering the following algorithm for binomial coefficient implementation in python.
```python
def factorial(n : int) -> int:
    # to be implemented

def binomial_coefficient(n : int, k : int):
    num = factorial(n)
    den = factorial(k) * factorial(n - k)
    return num // den
```
1. complete the factorial function using recursive methode.
2. profile the `binomial_coefficient`.
3. try to improve factorial function by using iterative method.
4. Use pyccel to accelerate the function.
5. compare the benchmark of the two version python vs pyccel.

### Completing the factorial function

In [41]:
def factorial(n : int) -> int:
    if n <= 0: 
        return 1
    else: 
        return n * factorial(n - 1)

In [45]:
def binomial_coefficient(n : int, k : int):
    num = factorial(n)
    den = factorial(k) * factorial(n - k)
    return num // den

### Profiling the binomial_coefficient

In [46]:
%timeit binomial_coefficient(10, 5)

2.42 µs ± 13.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [47]:
%lprun -f binomial_coefficient binomial_coefficient(10, 5)

###  Improving factorial function by using iterative method

In [48]:
from pyccel.decorators import types
from pyccel.epyccel import epyccel

In [49]:
@types( int )
def factorial_py(n):
    fact = 1
    for i in range(2, n+1):
        fact *= i
    return fact

### Pyccelize factorial_py

In [50]:
factorial_py = epyccel(factorial_py)

In [51]:
def binomial_coefficient_py(n, k):
    num = factorial_py(n)
    den = factorial_py(k) * factorial_py(n - k)
    return num // den

In [52]:
print("Time for binomial_coefficient using simple python")
%timeit binomial_coefficient(10, 5)
print("\n")
print("Time for binomial_coefficient_py with pyccel")

%timeit binomial_coefficient_py(10, 5)

Time for binomial_coefficient using simple python
2.49 µs ± 48 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


Time for binomial_coefficient_py with pyccel
461 ns ± 4.74 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


### Benchmark Comparison

**Python = 2.49 µs**<br/>
**Pyccel = 461 ns**

**<center>Thank you !</center>**